src/NLog/Internal/MruCache.cs

```diff
@@ -0,0 +1,185 @@
+// 
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution. 
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+using System.Collections.Generic;
+
+namespace NLog.Internal
+{
+    /// <summary>
+    /// Most-Recently-Used-Cache, that discards less frequently used items on overflow
+    /// </summary>
+    internal class MruCache<TKey, TValue>
+    {
+        private readonly Dictionary<TKey, MruItem> _dictionary;
+        private readonly int _maxCapacity;
+        private long _currentVersion;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="maxCapacity">Max number of items to hold before discarding</param>
+        public MruCache(int maxCapacity)
+        {
+            _maxCapacity = maxCapacity;
+            _dictionary = new Dictionary<TKey, MruItem>(_maxCapacity);
+            _currentVersion = 1;
+        }
+
+        /// <summary>
+        /// Insert item into caceh
+        /// </summary>
+        /// <param name="key">Dictionary Key</param>
+        /// <param name="value">Dictionary Value</param>
+        /// <returns>Key did not exist already</returns>
+        public bool TryAddValue(TKey key, TValue value)
+        {
+            lock (_dictionary)
+            {
+                MruItem item;
+                if (_dictionary.TryGetValue(key, out item))
+                {
+                    var version = _currentVersion;
+                    if (item.Version != version || !EqualityComparer<TValue>.Default.Equals(value, item.Value))
+                    {
+                        _dictionary[key] = new MruItem(value, version);
+                    }
+                    return false;   // Already exists
+                }
+                else
+                {
+                    var version = ++_currentVersion;
```

Comment:
I want "version" to have the updated value. I guess I could write:

`var version = _currentVersion++ + 1 ;`